[Wallace] iP

wallacepck opened the pull request:
### Mana

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Mana frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. download it from [release pending].
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```Java
public class Mana {
    public static void main(String[] args) {
        ...
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wallacepck committed:
Rename main class, logo

wallacepck committed:
Add greeting, reformat logo

wallacepck committed:
Alter greeting, add echo

wallacepck committed:
Add task list and task adding

wallacepck committed:
Add task class and new operations

Add task class
Add operations to add task, mark task as done/undone

wallacepck committed:
Add task subclasses, operations to add them

wallacepck committed:
Add general test, Add testmode to program

wallacepck committed:
Cleanup and refactor errors to exceptions

wallacepck committed:
Add delete operation

TeaMakesMePee commented:
really good

wallacepck committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

wallacepck committed:
Add gradle, Add GSON dep

wallacepck committed:
Move project files into proper main package

wallacepck committed:
Fix main class in gradle build script

wallacepck committed:
Fix wrong symbol for Event task

wallacepck committed:
Fix package name

wallacepck committed:
Add save/load to file functionality for task list

Add GSON serialiser for tasks
Add serialisation registrar
Add file manager to operate on savefile

wallacepck committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save/load to file functionality for task list

wallacepck committed:
Change Deadline and Event to use datetime literals

wallacepck committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Change Deadline and Event to use datetime literals

wallacepck committed:
Fix inconsistent date format in serializer

wallacepck committed:
Rename io package to better fit its purpose

wallacepck committed:
Add tasklist class

wallacepck committed:
Refactor commands, TaskList;  Add UserInterface

Commands are now created using the Builder pattern
User interaction is now standardised in UserInterface

wallacepck committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Refactor commands, TaskList;  Add UserInterface
  Add tasklist class

wallacepck committed:
Add tests

wallacepck committed:
Readd missing exit message

wallacepck committed:
Configure mainClassname for shadow

wallacepck committed:
Fix farewell message in wrong location

wallacepck committed:
Add immutable pair utility class

wallacepck committed:
Add find function

wallacepck committed:
Add javadocs for most classes

wallacepck committed:
Change imports to reflect code style standard

wallacepck committed:
Merge branch 'branch-A-JavaDoc'

wallacepck committed:
Merge branch 'branch-Level-9'

wallacepck committed:
Fix git tracking wrong directory due to rename

luna-ortus-cor commented on src/main/java/mana/Mana.java:
```diff
@@ -0,0 +1,73 @@
+                return;
+            }
+        } finally {
+            if (tasks == null) tasks = new TaskList();
```
```suggestion
            if (tasks == null) {
                tasks = new TaskList();
            }
```
to comply with the coding standard

luna-ortus-cor commented on src/main/java/mana/command/CommandParser.java:
```diff
@@ -0,0 +1,114 @@
+            throw new ManaException("No such command: %s", words[0]);
+        }
+    }
+}
```
Very clean code with good usage of ```BiFunction```.

wallacepck replied:
Thanks! Composition is king😎

luna-ortus-cor commented on src/main/java/mana/command/CommandTransformers.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Transformers transform arguments into valid Command parameters
+ */
+public class CommandTransformers {
```
```suggestion
public class CommandTransformer {
```
I think it would be good to use the singular instead of the plural here for consistency with the above ```CommandParser``` class

wallacepck replied:
I believe this is mistake on my part, I name the class `CommandParser`, but it additionally has the responsibility of storing `Command`s, which is bad practice. However, `CommandTransformers` *is* a holding class for various transformers, rather than it defining an object that is a "CommandTransformer".

luna-ortus-cor commented on src/main/java/mana/storage/TaskTypeAdapter.java:
```diff
@@ -0,0 +1,66 @@
+     * @param clazz The subclass of {@link Task}.
+     * @param typename The serialised name of {@code clazz}.
+     */
+    public static void register(Class<? extends Task> clazz, String typename) {
```
Is there a reason for naming the variable ```clazz``` instead of ```taskClass``` or similar?

wallacepck replied:
I'm used to naming `Class` parameters with limited scope as "clazz" to differentiate them from regular `Object`s, in line with `java` packages and many others. I do agree however it would be slightly clearer to name it verbosely.

luna-ortus-cor commented on src/main/java/mana/util/ImmutablePair.java:
```diff
@@ -0,0 +1,19 @@
+    public String toString() {
+        return String.format("Pair < %s , %s >", first, second);
+    }
+}
```
Good use of helper data structures for cleaner code.

luna-ortus-cor reviewed (commented):
Overall this is really well written and cleanly structured. Some minor suggestions only.

wallacepck committed:
Add checkstyle and refactor accordingly

Eun777 commented on src/main/java/mana/Mana.java:
Overall, good code structure. The variable names are clear and mostly follow the coding standard. Some additional JavaDocs and comments for this class would make it even better but good job overall!

Eun777 commented on src/test/java/mana/command/CommandTest.java:
```diff
@@ -0,0 +1,46 @@
+
+public class CommandTest {
+    @Test
+    public void action_noOp_success() {
```
Could be slightly improved with better naming of variables but otherwise, looks good.

Eun777 reviewed (commented):
Good structure and OOP overall. Some minor coding standard suggestions.

wallacepck committed:
Convert to javafx GUI, refactor UI accordingly

wallacepck committed:
Fix DelayedExit javadoc and handle illegal state

wallacepck committed:
Add assertions for nulls that would break program behaviour

Methods handling input and task finding are not nullsafe, and are not asserted.

Future changes could introduce scenarios whereby a null appears in these areas.

Adding assertions to these methods will guard against changes inadvertently introducing nulls, and allow developers reading the code to follow this assumption.

Introducing exceptions would be undesirable as disrupting the control flow makes no sense in these scenarios. Enforcing nullchecks has a similar issue, and would suggest that nulls are tolerable here when they are not.

wallacepck committed:
Refactor TaskList find method to use streams

Current TaskList find method uses an indexed for loop and multiple conditions to check for title match.

This is overtly verbose and confusing to navigate due to premature optimisation.

Changing the search logic to use streams results in cleaner and more readable code.

wallacepck committed:
Merge pull request #1 from wallacepck/branch-A-Assertions

Add assertions for nulls that would break program behaviour

wallacepck committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/mana/util/TaskList.java

wallacepck committed:
Merge pull request #2 from wallacepck/branch-A-Streams

Refactor TaskList find method to use streams

wallacepck committed:
Refactor Command class to remove arrowhead

wallacepck committed:
Merge branch 'master' into branch-A-CodeQuality

wallacepck committed:
Merge pull request #4 from wallacepck/branch-A-CodeQuality

Improve code quality

wallacepck committed:
Add command to load tasklist from a different file

wallacepck committed:
Fix timer zombie thread blocking program termination

wallacepck committed:
Add error handling when parameters are missing

wallacepck committed:
Handle case where parameter is used more than once

wallacepck committed:
Add prelim user documentation, Change window title

wallacepck committed:
Improve user docs using AB3 as reference

wallacepck committed:
Change readme to use program name

wallacepck committed:
Fix wrong command keyword in userdoc

wallacepck committed:
Add product screenshots

wallacepck committed:
Fix release build capitalisation

wallacepck committed:
Fix github pages
